tests/interfaces/test_autograd.py

```diff
@@ -475,3 +475,234 @@ def d_classical(a, b, method):
+            # differentiable quantum function
+            qml.templates.StronglyEntanglingLayers(weights, wires=[0, 1])
+            # non-differentiable quantum function
+            qml.templates.AngleEmbedding(data2, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        # differentiating the circuit wrt the weights
+        grad_fn = qml.grad(circuit)
+
+        # input weights
+        weights = qml.init.strong_ent_layers_normal(n_wires=2, n_layers=3)
+
+        # input data
+        data1 = qml.numpy.array([0, 1, 1, 0], requires_grad=False) / np.sqrt(2)
+        data2 = qml.numpy.array([1, 1], requires_grad=False)
+
+        res = grad_fn(data1, data2, weights)
+
+        # we do not check for correctness, just that the output
+        # is the correct shape
+        assert len(res) == 1
+        assert res[0].shape == weights.shape
+
+        # check that the parameter shift was only performed for the
+        # differentiable input, not the data input
+        assert spy.call_args[1]["wrt"] == set(data1.size + data2.size + np.arange(weights.size))
+
+    def test_multiple_differentiable_and_non_differentiable_parameters(self, mocker):
+        """Test that multiple differentiable and non-differentiable parameters
+        works as expected"""
+        spy = mocker.spy(qml.qnodes.JacobianQNode, "jacobian")
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev, interface="autograd")
+        def circuit(data1, weights1, data2, weights2):
+            qml.templates.AmplitudeEmbedding(data1, wires=[0, 1])
+            qml.templates.StronglyEntanglingLayers(weights1, wires=[0, 1])
+            qml.templates.AngleEmbedding(data2, wires=[0, 1])
+            qml.templates.StronglyEntanglingLayers(weights2, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        # differentiating the circuit wrt the weights
+        grad_fn = qml.grad(circuit)
+
+        # input weights
+        weights1 = qml.init.strong_ent_layers_normal(n_wires=2, n_layers=3)
+        weights2 = qml.init.strong_ent_layers_normal(n_wires=2, n_layers=4)
+
+        # input data
+        data1 = qml.numpy.array([0, 1, 1, 0], requires_grad=False) / np.sqrt(2)
+        data2 = qml.numpy.array([1, 1], requires_grad=False)
+
+        res = grad_fn(data1, weights1, data2, weights2)
+
+        # we do not check for correctness, just that the output
+        # is the correct shape
+        assert len(res) == 2
+        assert res[0].shape == weights1.shape
+        assert res[1].shape == weights2.shape
+
+        # check that the parameter shift was only performed for the
+        # differentiable input, not the data input
+        diff_variables = spy.call_args[1]["wrt"]
+        expected_diff_variables = set(data1.size + np.arange(weights1.size))
+        expected_diff_variables.update(data1.size + weights1.size + data2.size + np.arange(weights2.size))
+        assert diff_variables == expected_diff_variables
```

Comment:
I'm finding these `assert`s hard to easily understand

Reply:
I'll rewrite them explicitly :+1: